src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> previous;
+
+        private Node(Node<T> previous, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.previous = previous;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> newNode = new Node<>(null, value, null);
+            head = tail = newNode;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Wrong Index.");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index != 0 && index != size) {
+            Node<T> currentNode = findNode(index);
+            Node<T> addNode = new Node<>(null, value, currentNode);
+            addNode.previous = currentNode.previous;
+            currentNode.previous.next = addNode;
+            currentNode.previous = addNode;
+            size++;
+            return;
+        }
+        if (index == 0) {
+            Node<T> addNode = new Node<>(null, value, head);
+            head.previous = addNode;
+            head = addNode;
+            size++;
+            return;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = findNode(index);
+        T tempItem = currentNode.item;
+        currentNode.item = value;
+        return tempItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = findNode(index);
+        unlinkNode(currentNode);
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> searchNode = head;
+        while (searchNode != null) {
+            if (searchNode.item == object
+                    || (searchNode.item != null && searchNode.item.equals(object))) {
+                unlinkNode(searchNode);
+                return true;
+            }
+            searchNode = searchNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlinkNode(Node node) {
```

Comment:
think about how could we simplify this method. In each of the `if` statements you have a lot of repeating logic, think about how to combine it